ethcore/src/executive.rs

```diff
@@ -152,8 +153,13 @@ impl TransactOptions<trace::NoopTracer, trace::NoopVMTracer> {
 	}
 }
 
-pub fn executor(machine: &Machine, vm_factory: &Factory, params: &ActionParams) -> Box<vm::Vm> {
-	if machine.supports_wasm() && params.code.as_ref().map_or(false, |code| code.len() > 4 && &code[0..4] == WASM_MAGIC_NUMBER) {
+pub fn executor(
+	machine: &Machine,
+	vm_factory: &Factory,
+	params: &ActionParams,
+	blocknumber: BlockNumber,
+) -> Box<vm::Vm> {
+	if machine.supports_wasm(blocknumber) && params.code.as_ref().map_or(false, |code| code.len() > 4 && &code[0..4] == WASM_MAGIC_NUMBER) {
```

Comment:
I don't like this change at all. The `if` statement that we have right now is terrible and this change makes it even worse. This function/interface needs to be completely reworked

Reply:
Yeah, perhaps it should just go to `vm_factory`? This function seems to be another factory that delegates to `vm_factory` in case it's not wasm.

```rust
vm_factory.create(params.gas, params.code, blocknumber)
```

Reply:
I wanted to go this way, but last time I checked it was just not possible for some reason I can't recall. perhaps I should try again and find this reason again